### PR TITLE
Reduce perf gap between thread pool and omp

### PR DESCRIPTION
**Pull request by RandySheriffH**
Enabling thread pool of dispatching task asynchronously to remove perf bottleneck.
Test proved that the change brings general gains to models run on polaris:
Mobilenet(master vs fix):
image
https://lego-us.corp.microsoft.com/polaris/jobs/4c489e8f-f120-4658-825c-6cc40a9751cd
Prod_model11(master vs fix)
image
https://lego-us.corp.microsoft.com/polaris/jobs/9379ae34-776b-4a03-b694-3dd707ecb3b6
Resnet101(master vs fix)
image
https://lego-us.corp.microsoft.com/polaris/jobs/45f3d657-ad05-40eb-b2d6-7669b561c72b
Prod_model2(master vs fix)
image
https://lego-us.corp.microsoft.com/polaris/jobs/5fb6e249-bf56-4d81-8c15-d9b49f179544
Prod_mode12(master vs fix)
image
https://lego-us.corp.microsoft.com/polaris/jobs/d7ca3d05-2c94-42fe-9c30-f9cfed927f2b

### CODE REVIEW

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 1]
```diff
@@ -712,6 +712,7 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
         num_threads_(num_threads),
         allow_spinning_(allow_spinning),
         set_denormal_as_zero_(thread_options.set_denormal_as_zero),
+        enable_good_worker_hints_(thread_options.enable_good_worker_hints),
```
Comment (tlh20) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:715]: Having options like these are useful for our perf testing (avoiding the need to rebuild, upload new images to Polaris / Anubis / etc), but can we remove the flag prior to merging to the main branch.

Suggest we retain GoodWorkerHints in this PR, and then remove in the "sticky" PR.

Generally we want to keep a small number of configuration options, mapped to user-visible features such as the CPU vs perf trade-off in allow_spinning.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 2]
```diff
@@ -922,7 +923,7 @@ void EndParallelSectionInternal(PerThread &pt,
 
   // Wait for workers to exit ParLoopWorker
   auto tasks_to_wait_for = tasks_started - tasks_revoked;
-  while (ps.tasks_finished < tasks_to_wait_for) {
+  while (ps.tasks_finished.load(std::memory_order_acquire) < tasks_to_wait_for) {
```
Comment (tlh20) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:926]: Do we have evidence that we benefit from adding additional controls over memory ordering?  It adds complexity.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 3]
```diff
@@ -1078,29 +1080,107 @@ void RunInParallelSection(ThreadPoolParallelSection &ps,
...
-  profiler_.LogEndAndStart(ThreadPoolProfiler::DISTRIBUTION);
+  // index of thread that dispatch work to all other threads
+  int dispatch_q_idx = -1;
+  int dispatch_w_idx = -1;  // index of enqueued work
+  std::atomic<bool> dispatch_done{false};
+  std::atomic<bool> dispatch_work_done{false};
 
-  // Run work in the main thread
-  fn(0);
+  Task worker_fn = [&]() {
```
Comment (tlh20) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:1095]: Can we expand the [&] to be explicit about what is being captured.  A common bug is to implicitly capture local variables, and then access them once they are out of scope.  Making the captures explicit makes it easier to check that we are only handling safe cases (e.g., by-value capture from this scope, and capture of reference-typed variables).

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 4]
```diff
@@ -1078,29 +1080,107 @@ void RunInParallelSection(ThreadPoolParallelSection &ps,
...
+      }
+      return static_cast<int>(w_idx);
+    }
+  };
+
+  unsigned current_dop = static_cast<unsigned>(ps.tasks.size()) + 1;
+  unsigned extra_needed = n > current_dop ? n - current_dop : 0;
+  std::vector<unsigned> good_hints, alt_hints;
+
+  Task dispatch_task = [&]() {  // dispatch work to other threads
```
Comment (tlh20) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:1120]: Could you add some comments here on the overall approach, and timeline of operations in (1) the main thread, (2) the first thread running dispatch work, and (3) the additional threads helping in the loop.  Maybe ASCII-art :)

This is important for correctness -- if I am following the logic correctly, we need the main thread to remain in this function (hence for the dispatch thread to be OK accessing the local variables it captures).

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h [hunk 5]
```diff
@@ -1078,29 +1080,107 @@ void RunInParallelSection(ThreadPoolParallelSection &ps,
...
+      }
+    } else {
+      dispatch_q_idx = Rand(&pt->rand) % num_threads_;
+    }
+    dispatch_w_idx = enqueue_fn(dispatch_q_idx, dispatch_task, false);  // dispatch asynchronously
+    profiler_.LogEnd(ThreadPoolProfiler::DISTRIBUTION_ENQUEUE);
+  }
+
+  profiler_.LogEndAndStart(ThreadPoolProfiler::DISTRIBUTION);
+  fn(ps.worker_idx.fetch_add(1, std::memory_order_relaxed));  // run work in the main thread
```
Comment (tlh20) [inline include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h:1165]: Suggest running fn(0) here, and allocating IDs to other worker threads from 1 upwards.  That provides some additional cache locality (when ID X from one loop is tending to access common data with ID X from previous loops), and makes a slight simplification on the logic.

**Review by tlh20 — CHANGES_REQUESTED** [review]
Adding some detailed comments to the main changes here.  The overall approach looks good (assuming no surprises from perf tests).  We should clarify the algorithm via comments in the source code -- maybe ASCII art diagram of the timeline -- and I suggest we do not expose control over the GoodWorkerHint fn (given that we anticipate removing that entirely in the subsequent PR).

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 2 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: Change restored, will test with that separately.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 1 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: Reverted, let's do it in sticky PR.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 3 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: Done.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 4 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: Comments added ahead of function stating behaviors of main, dispatching and all other threads.

File: include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h (hunk 5 above)
Comment (RandySheriffH) [reply include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h]: Done.

**Review by tlh20 — CHANGES_REQUESTED** [review]
Thanks for the updates.  The logic here looks good now.  We need to move this to be in SummonWorkers though (so it applies in common across RunInParallel and RunInParallelSection).  I'd suggest:

* Rename SummonWorkers -> RunInParallelInternal (to mirror naming of other methods)

* Have RunInParallelInternal then run fn(0) in the caller.  [ I think this needs to be combined with the dispatch code, right, given the lifetimes of some of the variables captured in the lambdas. ]

Also, could we keep this PR retaining the GoodWorkerHints variant.  That will help with merging the "sticky" PR.

**Review by tlh20 — APPROVED** [review]

**Pull request merged**